### ISSUE TITLE
deleting a mixer rule takes a long time to take effective

### ISSUE THREAD

**Issue by linsun** [istio/old_mixer_repo#1451]
Applying seems fast enough but deleting can take 20 mins....   I'm doing the steps with mixer-rule-deny-label.yaml  and mixer-rule-deny-serviceaccount.yaml , see some details in this PR:  

https://github.com/istio/istio/pull/1159#issuecomment-337270116

**Comment by linsun** [comment]
https://istio.io/docs/tasks/security/basic-access-control.html#access-control-using-denials & https://istio.io/docs/tasks/security/secure-access-control.html#access-control-using-denials

**Comment by jmuk** [comment]
Lin, can you write down the steps to reproduce the way?  I did the following steps,

1. `% kubectl create -f install/kubernetes/istio-auth.yaml`
2. (some other addon installations, grafana, sourcegraph, prometheus, zipkin)
3. `% kubectl apply -f <(istioctl kube-inject -f samples/bookinfo/kube/bookinfo.yaml)`
4. see /productpage on the browser
5. `% istioctl create -f samples/bookinfo/kube/route-rule-reviews-v3.yaml`
6. `% istioctl create -f samples/bookinfo/kube/route-rule-reviews-test-v2.yaml`
7. reload /productpage on the browser, red stars appear
8. login as 'json', reload, black stars appear
9. logout
10. `% istioctl create -f samples/bookinfo/kube/mixer-rule-deny-label.yaml`
11. reload /productpage, stars disappear
12. `% istioctl delete -f samples/bookinfo/kube/mixer-rule-deny-label.yaml`
13. reload /productpage, red stars appear

I think your bug is need to wait for 20 mins before reload at step 13. Am I reading correctly?
It does not happen on me.  Similarly not reproducible with mixer-rule-deny-serviceaccount.yaml

**Comment by jmuk** [comment]
also wondering how mixer log looks like. Once a new operation on the rules (create/delete) arrives to mixer, mixer will print out logs like
```
I1019 19:00:34.787412       1 controller.go:207] Publishing 3 events
I1019 19:00:34.787673       1 controller.go:167] Published snapshot[8] with 0 rules, 0 handlers, previously 1 rules
I1019 19:00:34.787722       1 controller.go:550] cleanupResolver[7] handler table has 1 entries
I1019 19:00:34.787828       1 controller.go:559] closing denyproductpagehandler.denier.istio-system/&{{7 Not allowed []}}
```

**Comment by linsun** [comment]
yes, that is correct, for step 13, the removal of mixer-rule-deny-label.yaml doesn't take effective right away... I had to wait 20 mins in my case.  I also repeated the test with mixer-rule-deny-serviceaccount.yaml by following example here:  https://istio.io/docs/tasks/security/secure-access-control.html#access-control-using-denials and have same issue with deleting of the rule doesn
't take effective right away.

One thing tho, I tried with the namespace being removed as described in this patch:  https://github.com/istio/istio/pull/1159

**Comment by jmuk** [comment]
please let me know what the mixer log says; when does the "published snapshot" log appear? I think that it's delayed 20 minutes.

It is still unclear to me why that's delayed so long, maybe kubernetes system doesn't deliver the delete event to mixer?

What will happen when you do after step 12 (before 20 minutes wait)?
```sh
kubectl get rules.config.istio.io -o yaml --all-namespaces
```

This shows the kubernetes representation of the mixer rule; is it there?  Do there happen to be some finalizers?

**Comment by mandarjog** [comment]
@jmuk I am wondering if this is related to https://github.com/istio/mixer/issues/1373

**Comment by jmuk** [comment]
@mandarjog aw, I see. If that's the case, this is already fixed as 0.2.9?

**Comment by mandarjog** [comment]
yes, fixed.
